spock-specs/src/test/groovy/org/spockframework/smoke/Interceptors.groovy

```diff
@@ -155,64 +155,203 @@ class SubSpec extends SuperSpec {
+            assert false: 'currentIteration should not be set'
+          } catch (IllegalStateException ise) {
+            assert ise.message == 'Cannot request current iteration in @Shared context, or feature context'
+          }
+        }
+      }
+    }
+
+    static assertSpecMethodContext(IMethodInvocation invocation) {
+      invocation.with {
+        assert spec
+        assert !feature
+        assert !iteration
+        assert instance == sharedInstance
+        assert target == instance
+        assert method
+        assert method.reflection
+        assert method.name
+        instance.specificationContext.with {
+          assert currentSpec
+          try {
+            currentFeature
+            assert false: 'currentFeature should not be set'
+          } catch (IllegalStateException ise) {
+            assert ise.message == 'Cannot request current feature in @Shared context'
+          }
+          try {
+            currentIteration
+            assert false: 'currentIteration should not be set'
+          } catch (IllegalStateException ise) {
+            assert ise.message == 'Cannot request current iteration in @Shared context, or feature context'
+          }
+        }
+      }
+    }
+
+    static assertFeatureContext(IMethodInvocation invocation) {
+      invocation.with {
+        assert spec
+        assert feature
+        assert !iteration
+        assert instance == sharedInstance
+        assert target != instance
+        assert method
+        assert !method.reflection
+        assert !method.name
+        instance.specificationContext.with {
+          assert currentSpec
+          assert currentFeature
+          try {
+            currentIteration
+            assert false: 'currentIteration should not be set'
+          } catch (IllegalStateException ise) {
+            assert ise.message == 'Cannot request current iteration in @Shared context, or feature context'
+          }
+        }
+      }
+    }
+
+    static assertIterationContext(IMethodInvocation invocation) {
+      invocation.with {
+        assert spec
+        assert feature
+        assert iteration
+        assert instance != sharedInstance
+        assert target != instance
+        assert method
+        assert !method.reflection
+        assert !method.name
+        instance.specificationContext.with {
+          assert currentSpec
+          assert currentFeature
+          assert currentIteration
+        }
+      }
+    }
+
+    static assertIterationMethodContext(IMethodInvocation invocation) {
+      invocation.with {
+        assert spec
+        assert feature
+        assert iteration
+        assert instance != sharedInstance
+        assert target == instance
+        assert method
+        assert method.reflection
+        assert method.name
+        instance.specificationContext.with {
+          assert currentSpec
+          assert currentFeature
+          assert currentIteration
+        }
+      }
+    }
```

Comment:
WDYT about extracting common assertions to highlight the differences better.

```suggestion
    static assertSpecCommonContext(IMethodInvocation invocation) {
      invocation.with {
        assert spec
        assert !feature
        assert !iteration
        assert instance == sharedInstance
        assert method
        instance.specificationContext.with {
          assert currentSpec
          try {
            currentFeature
            assert false: 'currentFeature should not be set'
          } catch (IllegalStateException ise) {
            assert ise.message == 'Cannot request current feature in @Shared context'
          }
          try {
            currentIteration
            assert false: 'currentIteration should not be set'
          } catch (IllegalStateException ise) {
            assert ise.message == 'Cannot request current iteration in @Shared context, or feature context'
          }
        }
      }
    }
    static assertSpecContext(IMethodInvocation invocation) {
      assertSpecCommonContext(invocation)
      invocation.with {
        assert target != instance
        assert !method.reflection
        assert !method.name
      }
    }

    static assertSpecMethodContext(IMethodInvocation invocation) {
      assertSpecCommonContext(invocation)
      invocation.with {
        assert target == instance
        assert method.reflection
        assert method.name
      }
    }

    static assertFeatureCommonContext(IMethodInvocation invocation) {
      invocation.with {
        assert spec
        assert feature
        assert method
        instance.specificationContext.with {
          assert currentSpec
          assert currentFeature
        }
      }
    }

    static assertFeatureContext(IMethodInvocation invocation) {
      assertFeatureCommonContext(invocation)
      invocation.with {
        assert !iteration
        assert instance == sharedInstance
        assert target != instance
        assert !method.reflection
        assert !method.name
        instance.specificationContext.with {
          try {
            currentIteration
            assert false: 'currentIteration should not be set'
          } catch (IllegalStateException ise) {
            assert ise.message == 'Cannot request current iteration in @Shared context, or feature context'
          }
        }
      }
    }

    static assertIterationCommonContext(IMethodInvocation invocation) {
      assertFeatureCommonContext(invocation)
      invocation.with {
        assert iteration
        assert instance != sharedInstance
        instance.specificationContext.with {
          assert currentIteration
        }
      }
    }
    static assertIterationContext(IMethodInvocation invocation) {
      assertIterationCommonContext(invocation)
      invocation.with {
        assert instance != sharedInstance
        assert target != instance
        assert !method.reflection
        assert !method.name
      }
    }

    static assertIterationMethodContext(IMethodInvocation invocation) {
      assertIterationCommonContext(invocation)
      invocation.with {
        assert method.reflection
        assert method.name
      }
    }
```

Reply:
Yeah, good idea, thanks.
With the errors in your suggested changes fixed, that is. :-D